File: scripts/check-nesting.py
#!/usr/bin/env python3
"""Repo-wide nesting-depth lint for Python sources.

Walks the backend tree and reports every function whose maximum
``if`` / ``for`` / ``while`` / ``try`` / ``with`` / ``match`` nesting
depth exceeds the configured budget (default: 3 — i.e. at most three
levels of compound-statement nesting inside a single function body).

Sibling to ``scripts/check-file-lines.mjs``; runs as part of the
``backend`` pre-commit + CI pipeline.

Usage::

    python scripts/check-nesting.py            # check the default tree
    MAX_DEPTH=2 python scripts/check-nesting.py  # tighter override

Exit codes:

  0 — every scanned function is within budget
  1 — at least one function exceeds the budget; offenders are printed

Why a custom script rather than a ruff/flake8 plugin: ``ruff`` has no
direct "max nesting depth" rule (the closest is the cognitive-complexity
family, which conflates several signals).  This is intentionally one
narrow signal — easy to read, easy to override, easy to remove if a
better off-the-shelf option lands.
"""

from __future__ import annotations

import ast
import os
import sys
from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

# Roots we scan.  Add new top-level Python trees here.
SCAN_ROOTS = ("backend",)

# Directory names we never descend into.
SKIP_DIRECTORIES = frozenset(
    {
        "__pycache__",
        ".venv",
        "venv",
        "node_modules",
        ".git",
        "build",
        "dist",
        "alembic",
        "tests",  # tests can nest more freely (e.g. parametrized assertions)
        ".pytest_cache",
        ".mypy_cache",
        ".ruff_cache",
    }
)

# Compound statements that count as a level of nesting.
NESTING_NODES: tuple[type[ast.AST], ...] = (
    ast.If,
    ast.For,
    ast.AsyncFor,
    ast.While,
    ast.Try,
    ast.With,
    ast.AsyncWith,
    ast.Match,
)

DEFAULT_MAX_DEPTH = int(os.environ.get("MAX_DEPTH", "3"))

# Repo-relative ``path::function`` keys that are exempt from the
# nesting budget.  Pre-existing tech debt — each entry should be
# tracked in a follow-up bean and removed once the function is
# flattened.  Mirror of the EXEMPT_PATH_FRAGMENTS pattern in
# ``scripts/check-file-lines.mjs``.
#
# DO NOT add new entries here as a workaround.  New code must come in
# under the budget; if you genuinely cannot, raise it in review.
EXEMPT_FUNCTIONS: frozenset[str] = frozenset(
    {
        # TODO(pawrrtal): flatten Gemini provider stream + helper.
        # `make_gemini_stream_fn` and its inner `stream_fn` close over
        # 6 levels because the Google SDK's event surface is a switch
        # statement nested inside the streaming for-loop.  Extract
        # event-type handlers (text / function-call / safety) into
        # module-level helpers.
        "backend/app/core/providers/gemini_provider.py::make_gemini_stream_fn",
        "backend/app/core/providers/gemini_provider.py::stream_fn",
        "backend/app/core/providers/gemini_provider.py::stream",
        # TODO(pawrrtal): flatten the agent loop's event drain.
        "backend/app/core/agent_loop/loop.py::_run_loop",
        # TODO(pawrrtal): flatten the Claude SDK event translators.
        # All three walk a discriminated union from the Anthropic SDK
        # via repeated isinstance checks; would benefit from a small
        # dispatch table.
        "backend/app/core/providers/claude_provider.py::_events_from_assistant",
        "backend/app/core/providers/claude_provider.py::_events_from_message",
        "backend/app/core/providers/claude_provider.py::_tool_result_to_text",
    }
)


@dataclass(frozen=True)
class Offender:
    """A function whose deepest nesting level breached the budget."""

    path: Path
    function: str
    line: int
    depth: int


def _max_depth(node: ast.AST, current: int = 0) -> int:
    """Return the max compound-stmt nesting depth reachable from *node*.

    *current* is the depth at which *node* itself sits (so the body of
    a top-level function is depth 0; an ``if`` directly inside it
    pushes its body to depth 1; an ``if`` inside that ``if`` reaches
    depth 2; etc).
    """
    deepest = current
    for child in ast.iter_child_nodes(node):
        if isinstance(child, NESTING_NODES):
            child_depth = _max_depth(child, current + 1)
        else:
            child_depth = _max_depth(child, current)
        if child_depth > deepest:
            deepest = child_depth
    return deepest


def _walk_functions(tree: ast.AST) -> Iterator[ast.FunctionDef | ast.AsyncFunctionDef]:
    for node in ast.walk(tree):
        if isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef)):
            yield node


def check_file(path: Path, max_depth: int) -> list[Offender]:
    """Return every function in *path* whose max nesting exceeds *max_depth*."""
    try:
        source = path.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return []
    try:
        tree = ast.parse(source, filename=str(path))
    except SyntaxError:
        return []
    offenders: list[Offender] = []
    rel = path.relative_to(REPO_ROOT).as_posix()
    for func in _walk_functions(tree):
        depth = _max_depth(func, current=0)
        if depth <= max_depth:
            continue
        key = f"{rel}::{func.name}"
        if key in EXEMPT_FUNCTIONS:
            continue
        offenders.append(
            Offender(path=path, function=func.name, line=func.lineno, depth=depth)
        )
    return offenders


def _iter_python_files(root: Path) -> Iterator[Path]:
    for dirpath, dirnames, filenames in os.walk(root):
        # In-place mutation prunes the walk.
        dirnames[:] = [d for d in dirnames if d not in SKIP_DIRECTORIES]
        for name in filenames:
            if name.endswith(".py"):
                yield Path(dirpath) / name


def main(argv: list[str]) -> int:
    max_depth = DEFAULT_MAX_DEPTH
    if len(argv) > 1:
        try:
            max_depth = int(argv[1])
        except ValueError:
            print(f"check-nesting: invalid max depth: {argv[1]!r}", file=sys.stderr)
            return 2

    offenders: list[Offender] = []
    for root_name in SCAN_ROOTS:
        root = REPO_ROOT / root_name
        if not root.exists():
            continue
        for py_file in _iter_python_files(root):
            offenders.extend(check_file(py_file, max_depth))

    if not offenders:
        print(f"check-nesting: OK (no functions exceed depth {max_depth})")
        return 0

    offenders.sort(key=lambda o: (-o.depth, str(o.path), o.line))
    print(
        f"check-nesting: {len(offenders)} function(s) exceed depth {max_depth}:\n",
        file=sys.stderr,
    )
    for offender in offenders:
        rel = offender.path.relative_to(REPO_ROOT)
        print(
            f"  depth={offender.depth}  {rel}:{offender.line}  in {offender.function}()",
            file=sys.stderr,
        )
    print(
        "\nFlatten with guard clauses or extract helpers to bring each function under the budget.",
        file=sys.stderr,
    )
    return 1


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: scripts/check-nesting.mjs
#!/usr/bin/env node
/**
 * Repo-wide nesting-depth lint for TypeScript / TSX sources.
 *
 * Walks the frontend tree, parses each file with the official TypeScript
 * compiler API, and reports every function whose maximum control-flow
 * nesting depth exceeds the configured budget (default: 3 — at most three
 * levels of compound-statement nesting inside a single function body).
 *
 * Sibling to:
 *   - `scripts/check-file-lines.mjs`  (file-length budget, TS/TSX/PY)
 *   - `scripts/check-nesting.py`      (Python nesting, AST-based)
 *
 * Counted compound statements:
 *   - IfStatement
 *   - ForStatement / ForInStatement / ForOfStatement
 *   - WhileStatement / DoStatement
 *   - TryStatement (counted once — its catch/finally don't add depth)
 *   - SwitchStatement
 *   - WithStatement (just in case — depth budget should kill it on sight)
 *
 * NOT counted (different smells, addressed by other lints):
 *   - Ternaries (`a ? b : c`) — Biome covers complex expressions.
 *   - Logical short-circuit (`a && b`) — same.
 *   - Block statements (`{ ... }`) — they're a syntactic grouping, not a
 *     control-flow level.
 *   - Nested function / arrow / class declarations — those are a fresh
 *     scope and we recurse into them with depth reset to 0, so an inline
 *     callback inside a for-loop doesn't blow the budget on its own.
 *
 * Why a custom script rather than a Biome rule: Biome ships
 * `noExcessiveCognitiveComplexity` which conflates several signals
 * (nesting, branch count, recursion) into a single number.  This is
 * intentionally one narrow signal — easy to read, easy to override,
 * easy to remove if a better off-the-shelf option lands.
 *
 * Usage:
 *   node scripts/check-nesting.mjs              # default depth 3
 *   MAX_DEPTH=2 node scripts/check-nesting.mjs  # tighter
 *
 * Exit codes:
 *   0 — every scanned function is within budget
 *   1 — at least one function exceeds the budget; offenders are printed
 */

import { promises as fs } from 'node:fs';
import { createRequire } from 'node:module';
import path from 'node:path';
import process from 'node:process';

const require_ = createRequire(import.meta.url);

const REPO_ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), '..');

/** Hard ceiling — functions beyond this fail the check. */
const MAX_DEPTH = Number.parseInt(process.env.MAX_DEPTH ?? '3', 10);

/** Roots we scan. Add new top-level source trees here. */
const SCAN_ROOTS = ['frontend', 'backend-ts'];

/** Directory names we never descend into. */
const SKIP_DIRECTORIES = new Set([
	'node_modules',
	'.next',
	'dist',
	'build',
	'.venv',
	'__pycache__',
	'.beans',
	'.claude',
	'.cursor',
	'.agents',
	'.codex',
	'.opencode',
	'.crush',
	'.zed',
	'.vscode',
	'.sentrux',
	'.private',
	'alembic',
	'tests',
	'__tests__',
]);

/** File extensions we care about. */
const TARGET_EXTENSIONS = new Set(['.ts', '.tsx']);

/** Suffixes that are exempt (test/spec files can nest more freely). */
const EXEMPT_SUFFIXES = ['.test.ts', '.test.tsx', '.spec.ts', '.spec.tsx', '.d.ts'];

/**
 * Repo-relative `path::function` keys exempt from the budget.
 *
 * Pre-existing tech debt — each entry should be tracked in a follow-up
 * bean and removed once the function is flattened.  Mirror of the
 * `EXEMPT_FUNCTIONS` discipline in `scripts/check-nesting.py`.
 *
 * DO NOT add new entries here as a workaround.  New code must come in
 * under the budget; if you genuinely cannot, raise it in review.
 *
 * Seeded empty — populated on first CI run if any pre-existing
 * frontend code breaches the budget.
 */
const EXEMPT_FUNCTIONS = new Set([]);

/**
 * Repo-relative path fragments exempt from scanning entirely.
 *
 * Mirrors the same vendored / generated carve-outs from
 * `scripts/check-file-lines.mjs` — files we don't own can't be held
 * to our local style budget.
 */
const EXEMPT_PATH_FRAGMENTS = [
	'frontend/components/ui/', // shadcn-generated primitives
	'frontend/lib/react-dropdown/', // vendored sibling package
	'frontend/lib/react-overlay/', // vendored sibling package
	'frontend/lib/react-chat-composer/', // vendored sibling package
	'routeTree.gen.ts', // TanStack Router generated tree
];

const ts = require_('typescript');

/**
 * Compound-statement node kinds.  Hitting one of these inside a
 * function body increments the nesting depth by 1.
 */
const NESTING_KINDS = new Set([
	ts.SyntaxKind.IfStatement,
	ts.SyntaxKind.ForStatement,
	ts.SyntaxKind.ForInStatement,
	ts.SyntaxKind.ForOfStatement,
	ts.SyntaxKind.WhileStatement,
	ts.SyntaxKind.DoStatement,
	ts.SyntaxKind.TryStatement,
	ts.SyntaxKind.SwitchStatement,
	ts.SyntaxKind.WithStatement,
]);

/**
 * Function-like node kinds.  Entering one resets the depth to 0 so
 * an inline callback inside a deep for-loop doesn't inherit the
 * outer nesting (we'd flag the outer for-loop on its own merits).
 */
const FUNCTION_KINDS = new Set([
	ts.SyntaxKind.FunctionDeclaration,
	ts.SyntaxKind.FunctionExpression,
	ts.SyntaxKind.ArrowFunction,
	ts.SyntaxKind.MethodDeclaration,
	ts.SyntaxKind.GetAccessor,
	ts.SyntaxKind.SetAccessor,
	ts.SyntaxKind.Constructor,
]);

/** Best-effort name for a function-like node. */
function functionName(node, parent) {
	if (node.name?.escapedText) return String(node.name.escapedText);
	// Arrow / function expression assigned to a variable: `const foo = () => {}`.
	if (parent && parent.kind === ts.SyntaxKind.VariableDeclaration && parent.name) {
		return String(parent.name.escapedText ?? parent.name.getText?.() ?? '<anonymous>');
	}
	// Property assignment: `{ foo: () => {} }`.
	if (parent && parent.kind === ts.SyntaxKind.PropertyAssignment && parent.name) {
		return String(parent.name.escapedText ?? parent.name.getText?.() ?? '<anonymous>');
	}
	if (node.kind === ts.SyntaxKind.Constructor) return 'constructor';
	return '<anonymous>';
}

/**
 * Recursively walk `node`, tracking the deepest nesting reached
 * inside the current function body.  When we cross into a new
 * function-like node we reset the depth and record that function's
 * deepest reach independently.
 */
function analyse(sourceFile) {
	const offenders = [];

	function walk(node, parent, currentDepth, currentFunction) {
		if (FUNCTION_KINDS.has(node.kind)) {
			const name = functionName(node, parent);
			const { line } = sourceFile.getLineAndCharacterOfPosition(node.getStart());
			const fnState = { name, line: line + 1, deepest: 0 };
			ts.forEachChild(node, (child) => walk(child, node, 0, fnState));
			if (fnState.deepest > MAX_DEPTH) {
				offenders.push(fnState);
			}
			return;
		}

		let nextDepth = currentDepth;
		if (NESTING_KINDS.has(node.kind)) {
			nextDepth = currentDepth + 1;
			if (currentFunction && nextDepth > currentFunction.deepest) {
				currentFunction.deepest = nextDepth;
			}
		}

		ts.forEachChild(node, (child) => walk(child, node, nextDepth, currentFunction));
	}

	ts.forEachChild(sourceFile, (child) => walk(child, sourceFile, 0, null));
	return offenders;
}

/** Recursively yield every TS/TSX source file under `dir` we should check. */
async function* walkDir(dir) {
	let entries;
	try {
		entries = await fs.readdir(dir, { withFileTypes: true });
	} catch {
		return;
	}
	for (const entry of entries) {
		if (entry.isDirectory()) {
			if (SKIP_DIRECTORIES.has(entry.name)) continue;
			yield* walkDir(path.join(dir, entry.name));
			continue;
		}
		if (!entry.isFile()) continue;
		const ext = path.extname(entry.name);
		if (!TARGET_EXTENSIONS.has(ext)) continue;
		if (EXEMPT_SUFFIXES.some((suffix) => entry.name.endsWith(suffix))) continue;
		yield path.join(dir, entry.name);
	}
}

async function checkFile(filePath, repoRelative) {
	const source = await fs.readFile(filePath, 'utf8');
	const scriptKind = filePath.endsWith('.tsx') ? ts.ScriptKind.TSX : ts.ScriptKind.TS;
	const sourceFile = ts.createSourceFile(
		filePath,
		source,
		ts.ScriptTarget.Latest,
		/* setParentNodes */ true,
		scriptKind
	);
	const offenders = analyse(sourceFile);
	return offenders
		.filter((o) => !EXEMPT_FUNCTIONS.has(`${repoRelative}::${o.name}`))
		.map((o) => ({ path: repoRelative, ...o }));
}

async function main() {
	const all = [];
	for (const root of SCAN_ROOTS) {
		const absoluteRoot = path.join(REPO_ROOT, root);
		try {
			await fs.access(absoluteRoot);
		} catch {
			continue;
		}
		for await (const filePath of walkDir(absoluteRoot)) {
			const relative = path.relative(REPO_ROOT, filePath);
			if (EXEMPT_PATH_FRAGMENTS.some((fragment) => relative.includes(fragment))) {
				continue;
			}
			const offenders = await checkFile(filePath, relative);
			all.push(...offenders);
		}
	}

	if (all.length === 0) {
		console.log(`check-nesting: OK (no functions exceed depth ${MAX_DEPTH})`);
		return;
	}

	all.sort((a, b) => b.deepest - a.deepest || a.path.localeCompare(b.path) || a.line - b.line);
	console.error(`check-nesting: ${all.length} function(s) exceed depth ${MAX_DEPTH}:\n`);
	for (const o of all) {
		console.error(`  depth=${o.deepest}  ${o.path}:${o.line}  in ${o.name}()`);
	}
	console.error(
		'\nFlatten with guard clauses or extract helpers to bring each function under the budget.'
	);
	process.exitCode = 1;
}

await main();
